Create migration guide for ASP.NET applications (#2864)
microsoft/ApplicationInsights-dotnet: docs/migration/guide_aspnet.md

# Guide for Asp.Net applications

## Intro

> The following guide details how to remove Application Insights from an ASP.NET web app and how to set up the OpenTelemetry SDK.

Reviewer: ```suggestion
The following guide details how to migrate from Application Insights SDK to OpenTelemetry based solution for an ASP.NET web application.
```

## Prerequisites

- An ASP.NET web application already instrumented with Application Insights

> - A actively supported version of .NET (link)

Reviewer: Be specific as ASP.NET is only for .NET Framework, so .NET462 or higher only.

## Steps to Migrate

### Step 1: Remove Application Insights SDK

> When you first added ApplicationInsights to your project, the SDK would have added a config file and made some edits to the web.config.

Reviewer: https://learn.microsoft.com/en-us/azure/azure-monitor/app/asp-net#how-can-i-uninstall-the-sdk

    ```xml
    <configuration>
      <system.web>
        <httpModules>
          <add name="TelemetryCorrelationHttpModule" type="Microsoft.AspNet.TelemetryCorrelation.TelemetryCorrelationHttpModule, Microsoft.AspNet.TelemetryCorrelation" />
          <add name="ApplicationInsightsWebTracking" type="Microsoft.ApplicationInsights.Web.ApplicationInsightsHttpModule, Microsoft.AI.Web" />
        </httpModules>
      </system.web>
      <system.webServer>
        <modules>
          <remove name="TelemetryCorrelationHttpModule" />
          <add name="TelemetryCorrelationHttpModule" type="Microsoft.AspNet.TelemetryCorrelation.TelemetryCorrelationHttpModule, Microsoft.AspNet.TelemetryCorrelation" preCondition="managedHandler" />
          <remove name="ApplicationInsightsWebTracking" />
          <add name="ApplicationInsightsWebTracking" type="Microsoft.ApplicationInsights.Web.ApplicationInsightsHttpModule, Microsoft.AI.Web" preCondition="managedHandler" />
        </modules>
      </system.webServer>
    </configuration>
    ```

### Step 1: Install the OpenTelemetry SDK and Enable at Application Startup

> The OpenTelemery SDK must be configured at application startup. This is typically done in the Global.asax.cs.

Reviewer: ```suggestion
The OpenTelemery SDK must be configured at application startup. This is typically done in the `Global.asax.cs`.
```

OpenTelemetry has a concept of three signals; Traces (Requests and Dependencies), Metrics, and Logs.
Each of these signals will need to be configured as part of your application startup.

#### Global.asax.cs

TODO: NEED A SAMPLE THAT INCLUDES LOGGING.

```csharp
using OpenTelemetry;
using OpenTelemetry.Metrics;
using OpenTelemetry.Trace;

public class MvcApplication : System.Web.HttpApplication
{
    private TracerProvider? tracerProvider;
    private MeterProvider? meterProvider;

    protected void Application_Start()
    {
        this.tracerProvider = Sdk.CreateTracerProviderBuilder()
            .Build();

        this.meterProvider = Sdk.CreateMeterProviderBuilder()
            .Build();
    }

    protected void Application_End()
    {
        this.tracerProvider?.Dispose();
        this.meterProvider?.Dispose();
    }
}

> ```

Reviewer: i'd recommend what happens at the end of this step - in this case nothing happens, but make it very clear that this step is just a prep-step only, and no telemetry will be collected at this point.

For more examples, see the following guides:
- OpenTelemetry SDK's getting started guide: https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry
- ASP.NET example project: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/examples/AspNet/Global.asax.cs

> ### Step 1: Configure Instrumentation Libraries.

Reviewer: this is step2 right?

Author: I'm saving this for last. Right now, all the Steps are listed as "1". I did this intentionally so I can focus on the content without needing to re-number every section when making edits.
